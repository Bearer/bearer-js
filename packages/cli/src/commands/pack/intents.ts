import * as archiver from 'archiver'
import * as fs from 'fs-extra'
import * as path from 'path'

import BaseCommand from '../../base-command'
import { RequireScenarioFolder } from '../../utils/decorators'
import prepareConfig from '../../utils/prepare-config'

const CONFIG_FILE = 'bearer.config.json'
const HANDLER_NAME = 'index'
const HANDLER_NAME_WITH_EXT = [HANDLER_NAME, 'js'].join('.')

export default class PackIntents extends BaseCommand {
  static description = 'Pack scenario intents'
  static hidden = true
  static flags = {
    ...BaseCommand.flags
  }

  static args = [{ name: 'ARCHIVE_PATH', required: true }]

  @RequireScenarioFolder()
  async run() {
    // we are assuming prepare and build have been run previously
    const { args } = this.parse(PackIntents)
    const { ARCHIVE_PATH } = args
    const target = path.resolve(ARCHIVE_PATH)
    this.debug(`Zipping to: ${target}`)

    const output = fs.createWriteStream(target)
    const archive = archiver('zip', {
      zlib: { level: 9 }
    })
    const archiveEntries: Array<archiver.EntryData> = []
    archive.pipe(output)

    const archived = new Promise<Array<archiver.EntryData>>((resolve, reject) => {
      output.on('close', () => {
        resolve(archiveEntries)
      })
      archive.on('error', (err: any) => {
        reject(err)
      })

      archive.on('entry', (entry: archiver.EntryData) => {
        if (!entry.stats || !entry.stats.isDirectory()) {
          archiveEntries.push(entry)
        }
      })

      archive.on('warning', (err: any) => {
        this.debug(err)
      })
    })

    try {
      // add intents
      archive.directory(this.locator.buildIntentsDir, false)
      // add CONFIG
      const intents = await this.retrieveIntents()

      const { config, handlers } = buildLambdaDefinitions(intents)
      const finalConfig = await this.retrieveAuthConfig(config)

      archive.append(JSON.stringify(finalConfig, null, 2), { name: CONFIG_FILE })

      // add handlers
      this.debug(`Generated config: ${JSON.stringify(config, null, 2)}`)
      archive.append(handlers, { name: HANDLER_NAME_WITH_EXT })

      // ZIP
      archive.finalize()

      const entries = await archived

      this.debug(`Zip content: ${entries.length} files\n  * ${entries.map(e => e.name).join('\n  * ')}`)
      this.success('Successfully generated lambda package')
      // log files added to zip
    } catch (e) {
      this.error(e)
    }
  }

  async retrieveAuthConfig(config: { intents: Record<string, string>[] }): Promise<TBearerConfig> {
    // generate config
    const content = await fs.readFile(this.locator.authConfigPath, { encoding: 'utf8' })
    return {
      ...config,
      auth: JSON.parse(content)
    }
  }

  // TODO: rewrite this using TS AST
  async retrieveIntents(): Promise<TIntentNames> {
    try {
      const config = await prepareConfig(
        this.locator.authConfigPath,
        this.bearerConfig.scenarioUuid,
        this.locator.srcIntentsDir
      )
      return config.intents
    } catch (e) {
      throw e
    }
  }
}

type TIntentNames = string[]

type TBearerConfig = {
  intents: Record<string, string>[]
  auth?: any
}

type TLambdaDefinition = {
  config: {
    intents: Record<string, string>[]
  }
  handlers: string
}

export function buildLambdaDefinitions(intents: TIntentNames): TLambdaDefinition {
  return {
    config: {
      intents: intents.reduce(
        (acc, intentName) => {
          acc.push({ [intentName]: [HANDLER_NAME, intentName].join('.') })
          return acc
        },
        [] as Record<string, string>[]
      )
    },
    handlers: buildLambdaIndex(intents)
  }
}

function buildLambdaIndex(intents: TIntentNames): string {
  return intents
    .map((intent, index) => {
      const intentConstName = `intent${index}`
      const intentInstance = `${intentConstName}Instance`
      return `const ${intentConstName} = require("./dist/${intent}").default;
const ${intentInstance} = new ${intentConstName}()
module.exports['${intent}'] = ${intentInstance}.intentType.intent(${intentInstance}.action);
`
    })
    .join('\n')
}
