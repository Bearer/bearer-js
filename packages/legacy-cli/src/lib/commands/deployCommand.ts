import { deployScenario, IDeployOptions } from '../deployScenario'
import Locator from '../locationProvider'

const deploy = (emitter, config, locator: Locator) => async ({ viewsOnly = false, intentsOnly = false }) => {
  emitter.emit('deploy:started')
  // Always true?
  if (!locator.scenarioRoot) {
    emitter.emit('rootPath:doesntExist')
    process.exit(1)
  }

  const { hasScenarioLinked, scenarioId } = config
  if (!hasScenarioLinked) {
    emitter.emit('scenarioUuid:missing')
    process.exit(1)
  }

  if (
    !(
      config.bearerConfig &&
      config.bearerConfig.authorization &&
      config.bearerConfig.authorization.AuthenticationResult &&
      config.bearerConfig.authorization.AuthenticationResult.IdToken
    )
  ) {
    emitter.emit('user:notAuthenticated')
    process.exit(1)
  }

  const deployOptions: IDeployOptions = { noViews: intentsOnly, noIntents: viewsOnly }

  try {
    await deployScenario(deployOptions, emitter, config, locator)
    const setupUrl = `${config.DeveloperPortalUrl}scenarios/${config.scenarioUuid}/preview`

    emitter.emit('deploy:finished', {
      scenarioId,
      setupUrl
    })
  } catch (error) {
    emitter.emit('deploy:failed', {
      error
    })
  }
}

export function useWith(program, emitter, config, locator): void {
  program
    .command('deploy')
    .description(
      `Build a scenario package.
$ bearer deploy
`
    )
    .option('-v, --views-only', 'Deploy views only')
    .option('-i, --intents-only', 'Deploy intents only')
    .action(deploy(emitter, config, locator))
}
