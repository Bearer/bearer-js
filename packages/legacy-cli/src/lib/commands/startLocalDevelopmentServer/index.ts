import * as getPort from 'get-port'
import * as Router from 'koa-router'
import * as cosmiconfig from 'cosmiconfig'
import * as Logger from 'koa-logger'
import * as chokidar from 'chokidar'

import { transpileIntents } from '../../buildArtifact'
import server = require('./server')
import Storage from './storage'
import auth from './auth'
import LocationProvider from '../../locationProvider'
import { Config } from '../../types'

function requireUncached(module) {
  delete require.cache[require.resolve(module)]
  return require(module)
}

export default function startLocalDevelopmentServer(
  emitter,
  config: Config,
  locator: LocationProvider,
  logs: boolean = true
) {
  const rootLevel = locator.scenarioRoot

  const LOCAL_DEV_CONFIGURATION = 'dev'
  const explorer = cosmiconfig(LOCAL_DEV_CONFIGURATION, {
    searchPlaces: [`config.${LOCAL_DEV_CONFIGURATION}.js`]
  })
  const router = new Router({ prefix: '/api/v1/' })

  return new Promise(async (resolve, reject) => {
    try {
      const { config: devIntentsContext = {} } = (await explorer.search(rootLevel)) || {}
      const distPath = locator.buildIntentsResourcePath('dist')

      async function refreshIntents() {
        try {
          emitter.emit('start:localServer:generatingIntents:start')
          await transpileIntents(locator.srcIntentsDir, distPath)
          emitter.emit('start:localServer:generatingIntents:stop')
        } catch (error) {
          emitter.emit('start:localServer:generatingIntents:failed', { error })
        }
      }
      await refreshIntents()

      chokidar
        .watch('.', {
          ignored: /(^|[\/\\])\../,
          cwd: locator.srcIntentsDir,
          ignoreInitial: true,
          persistent: true,
          followSymlinks: false
        })
        .on('add', refreshIntents)
        .on('change', refreshIntents)

      const port = await getPort({ port: 3000 })
      const bearerBaseURL = `http://localhost:${port}/`
      router.all(
        `${config.scenarioUuid}/:intentName`,
        (ctx, next) =>
          new Promise((resolve, reject) => {
            try {
              const intent = requireUncached(`${distPath}/${ctx.params.intentName}`).default
              intent.intentType.intent(intent.action)(
                {
                  context: {
                    ...devIntentsContext.global,
                    ...devIntentsContext[ctx.params.intentName],
                    bearerBaseURL
                  },
                  queryStringParameters: ctx.query,
                  body: JSON.stringify(ctx.request.body)
                },
                {},
                (_err, datum) => {
                  ctx.intentDatum = datum
                  next()
                  resolve()
                }
              )
            } catch (e) {
              reject({ error: e.toString() })
            }
          }),
        ctx => ctx.ok(ctx.intentDatum)
      )

      const storage = Storage()
      if (logs) {
        server.use(Logger())
      }
      server.use(storage.routes())
      server.use(storage.allowedMethods())
      server.use(auth.routes())
      server.use(auth.allowedMethods())
      server.use(router.routes())
      server.use(router.allowedMethods())

      server.listen(port, () => {
        emitter.emit('start:localServer:start', { port })
        emitter.emit('start:localServer:endpoints', {
          endpoints: [...storage.stack, ...auth.stack, ...router.stack]
        })
      })

      resolve(bearerBaseURL)
    } catch (e) {
      reject(e)
    }
  })
}
