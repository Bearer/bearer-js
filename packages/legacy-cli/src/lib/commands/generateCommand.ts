import * as intents from '@bearer/intents'
import * as templates from '@bearer/templates'
import * as Case from 'case'
import * as copy from 'copy-template-dir'
import * as inquirer from 'inquirer'
import * as path from 'path'

import Locator from '../locationProvider'

import { generateSetup } from './generate'

const INTENT = 'intent'
const COMPONENT = 'component'

const enum Components {
  BLANK = 'blank',
  COLLECTION = 'collection',
  ROOT = 'root'
}

const generate = (emitter, {}, locator: Locator) => async env => {
  const { scenarioRoot } = locator
  if (!scenarioRoot) {
    emitter.emit('rootPath:doesntExist')
    process.exit(1)
  }

  if (env.setup) {
    return generateSetup({ emitter, locator })
  }

  if (env.blankComponent && typeof env.blankComponent === 'string') {
    return generateComponent({ emitter, locator, name: env.blankComponent, type: Components.BLANK })
  }

  if (env.blankComponent) {
    return generateComponent({ emitter, locator, type: Components.BLANK })
  }

  if (env.collectionComponent && typeof env.collectionComponent === 'string') {
    return generateComponent({ emitter, locator, name: env.collectionComponent, type: Components.COLLECTION })
  }

  if (env.collectionComponent) {
    return generateComponent({ emitter, locator, type: Components.COLLECTION })
  }

  if (env.rootGroup && typeof env.rootGroup === 'string') {
    return generateComponent({ emitter, locator, name: env.rootGroup, type: Components.ROOT })
  }

  if (env.rootGroup) {
    return generateComponent({ emitter, locator, type: Components.ROOT })
  }

  const { template } = await inquirer.prompt([
    {
      message: 'What do you want to generate',
      type: 'list',
      name: 'template',
      choices: [{ name: 'Intent', value: INTENT }, { name: 'Component', value: COMPONENT }]
    }
  ])

  const params = { emitter, locator }

  switch (template) {
    case INTENT:
      await generateIntent(params)
      break
    case COMPONENT:
      await generateComponent(params)
      break
    default:
  }
}

type TgenerateComponent = { locator: Locator; emitter: any; name?: string; type?: string }
async function generateComponent({ emitter, locator, name, type }: TgenerateComponent) {
  // Ask for type if not present
  if (!type) {
    type = await askComponentType()
  }

  // Ask for name if not present
  if (!name) {
    name = await askForName()
  }

  const inDir = path.join(__dirname, 'templates/generate', `${type}Component`)
  const outDir = type === Components.ROOT ? locator.srcViewsDir : locator.srcViewsDirResource('components')

  copy(inDir, outDir, getComponentVars(name, require(locator.authConfigPath)), (err, createdFiles) => {
    if (err) throw err
    createdFiles.forEach(filePath => emitter.emit('generateView:fileGenerated', filePath))
  })
}

async function generateIntent({ emitter, locator }: { emitter: any; locator: Locator }) {
  const intentType = await askIntentType()
  const name = await askForName()
  const authConfig = require(locator.authConfigPath)
  const inDir = path.join(__dirname, 'templates/generate/intent')
  const outDir = locator.srcIntentsDir

  copy(inDir, outDir, getIntentVars(name, intentType, authConfig), (err, createdFiles) => {
    if (err) throw err
    createdFiles.forEach(filePath => emitter.emit('generateTemplate:fileGenerated', filePath))
  })
}

/**
 * Helpers
 */

export function getIntentChoices(): Array<{ name: string; value: any }> {
  const filteredChoices = (intents: Record<string, any>, propertyFlag) =>
    Object.keys(intents)
      .filter(intent => intents[intent][propertyFlag])
      .map(intent => ({
        name: intents[intent].display,
        value: intent
      }))
      .sort((a, b) => (a.name > b.name ? 1 : -1))

  return [
    ...filteredChoices(intents, 'isGlobalIntent'),
    new inquirer.Separator(),
    ...filteredChoices(intents, 'isStateIntent')
  ]
}

function getActionExample(intentType, authType): string {
  return templates[authType][intentType]
}

async function askIntentType(): Promise<string> {
  const { intentType } = await inquirer.prompt([
    {
      message: 'What type of intent do you want to generate',
      type: 'list',
      name: 'intentType',
      choices: getIntentChoices()
    }
  ])
  return intentType
}

async function askComponentType(): Promise<Components> {
  const typePrompt = await inquirer.prompt([
    {
      message: 'What type of component do you want to generate',
      type: 'list',
      name: 'type',
      choices: [
        { name: 'Blank', value: Components.BLANK },
        { name: 'Collection', value: Components.COLLECTION },
        new inquirer.Separator(),
        { name: 'Root Group', value: Components.ROOT }
      ]
    }
  ])
  return typePrompt.type
}

export function getComponentVars(name: string, authConfig: { authType: string }) {
  const componentName = Case.pascal(name)
  return {
    fileName: name,
    componentName,
    componentClassName: componentName, // it gives more meaning within templates
    componentTagName: Case.kebab(componentName),
    groupName: Case.kebab(componentName),
    withAuthScreen: authConfig.authType !== 'noAuth' ? '<bearer-navigator-auth-screen />' : null
  }
}

export function getIntentVars(name: string, intentType: string, authConfig: { authType: string }) {
  const actionExample = getActionExample(intentType, authConfig.authType)
  return {
    fileName: name,
    intentName: name,
    intentClassName: Case.pascal(name),
    authType: authConfig.authType,
    intentType,
    actionExample,
    callbackType: `T${intentType}Callback`
  }
}

async function askForName() {
  const { name } = await inquirer.prompt([{ message: 'Give it a name', type: 'input', name: 'name' }])
  return name.trim()
}

/**
 * Command
 */
export function useWith(program, emitter, config, locator): void {
  program
    .command('generate')
    .alias('g')
    .description(
      `Generate intent or component.
    $ bearer generate
  `
    )
    // .option('-t, --type <intentType>', 'Intent type.')
    .option('--blank-component [name]', 'generate blank component')
    .option('--collection-component [name]', 'generate collection component')
    .option('--root-group [name]', 'generate root components group')
    .option('--setup', 'generate setup file')
    .action(generate(emitter, config, locator))
}
