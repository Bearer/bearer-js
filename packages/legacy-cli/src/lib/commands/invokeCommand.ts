import axios from 'axios'
import * as cosmiconfig from 'cosmiconfig'

import Locator from '../locationProvider'
import startLocalDevelopmentServer from './startLocalDevelopmentServer'

type IConfig = {
  params?: {}
  body?: {}
}

export const invoke = (emitter, config, locator: Locator) => async (func, cmd) => {
  const { path } = cmd
  const { buid } = config

  let fileData: IConfig = {}
  if (path) {
    const explorer = cosmiconfig(path, {
      searchPlaces: [path]
    })
    const { config = {} } = (await explorer.search(locator.integrationRootResourcePath.toString())) || {}
    fileData = config
  }
  const { params = {}, body = {} } = fileData
  try {
    const integrationHostURL = await startLocalDevelopmentServer(emitter, config, locator, { force: true })

    const client = axios.create({ baseURL: `${integrationHostURL}api/v4/backend/functions`, timeout: 5000 })

    const { data } = await client.post(`${buid}/${func}`, body, { params })
    // used by cli: do not remove
    console.log(JSON.stringify(data, null, 2))
    process.exit(0)
  } catch (e) {
    if (e.response) {
      console.log(JSON.stringify(e.response.data, null, 2))
      process.exit(0)
    } else {
      console.log(e.toString())
      process.exit(1)
    }
  }
}

export function useWith(program, emitter, config, locator: Locator) {
  program
    .command('invoke <func>')
    .option('-p, --path <path>')
    .description(
      `invoke Function locally.
  $ bearer invoke <FunctionName>
  $ bearer invoke <FunctionName> -p tests/func.json
`
    )
    .action(invoke(emitter, config, locator))
}
