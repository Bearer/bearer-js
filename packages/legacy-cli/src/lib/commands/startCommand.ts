const path = require('path')
const fs = require('fs-extra')
const copy = require('copy-template-dir')
const Case = require('case')
const chokidar = require('chokidar')
const { spawn, execSync } = require('child_process')

import startLocalDevelopmentServer from './startLocalDevelopmentServer'
import Locator from '../locationProvider'
import { generateSetup } from './generate'

function watchNonTSFiles(watchedPath, destPath): Promise<any> {
  return new Promise((resolve, _reject) => {
    function callback(error) {
      if (error) {
        console.log('error', error)
      }
    }
    const watcher = chokidar.watch(watchedPath + '/**', {
      ignored: /\.tsx?$/,
      persistent: true,
      followSymlinks: false
    })

    watcher.on('ready', () => {
      resolve(watcher)
    })

    watcher.on('all', (event, filePath) => {
      const relativePath = filePath.replace(watchedPath, '')
      const targetPath = path.join(destPath, relativePath)
      // Creating symlink
      if (event == 'add') {
        console.log('creating symlink', filePath, targetPath)
        fs.ensureSymlink(filePath, targetPath, callback)
      }
      // // Deleting symlink
      if (event == 'unlink') {
        console.log('deleting symlink')
        fs.unlink(targetPath, err => {
          if (err) throw err
          console.log(targetPath + ' was deleted')
        })
      }
    })
  })
}

export function prepare(emitter, config, locator: Locator) {
  return async (
    { install = true, watchMode = true } = {
      install: true,
      watchMode: true
    }
  ) => {
    try {
      const {
        scenarioConfig: { scenarioTitle }
      } = config
      const { buildViewsDir, buildViewsComponentsDir, srcViewsDir, scenarioRoot } = locator

      // Create hidden folder
      emitter.emit('start:prepare:buildFolder')
      if (!fs.existsSync(buildViewsDir)) {
        fs.mkdirpSync(buildViewsDir)
      }
      fs.emptyDirSync(buildViewsDir)

      if (!fs.existsSync(buildViewsComponentsDir)) {
        fs.mkdirpSync(buildViewsComponentsDir)
      }

      // Symlink node_modules
      emitter.emit('start:symlinkNodeModules')
      createEvenIfItExists(
        locator.scenarioRootResourcePath('node_modules'),
        locator.buildViewsResourcePath('node_modules')
      )

      // symlink package.json
      emitter.emit('start:symlinkPackage')

      createEvenIfItExists(
        locator.scenarioRootResourcePath('package.json'),
        locator.buildViewsResourcePath('package.json')
      )

      // Copy stencil.config.json
      emitter.emit('start:prepare:stencilConfig')

      const vars = {
        componentTagName: Case.kebab(scenarioTitle)
      }
      const inDir = path.join(__dirname, 'templates', 'start')
      await new Promise((resolve, reject) => {
        copy(inDir, buildViewsDir, vars, (err, createdFiles) => {
          if (err) reject(err)
          createdFiles && createdFiles.forEach(filePath => emitter.emit('start:prepare:copyFile', filePath))
          resolve()
        })
      })

      createEvenIfItExists(
        locator.buildViewsResourcePath('global'),
        path.join(locator.buildViewsComponentsDir, 'global')
      )

      await generateSetup({ emitter, locator })

      // Link non TS files
      const watcher = await watchNonTSFiles(srcViewsDir, buildViewsComponentsDir)

      if (!watchMode) {
        watcher.close()
      }

      if (install) {
        emitter.emit('start:prepare:installingDependencies')
        execSync(`${config.command} install`, { cwd: scenarioRoot })
      }

      return {
        rootLevel: scenarioRoot,
        buildDirectory: buildViewsDir,
        viewsDirectory: srcViewsDir
      }
    } catch (error) {
      emitter.emit('start:prepare:failed', { error })
      throw error
    }
  }
}

const ensureSetupComponents = (emitter, locator) => {
  generateSetup({ emitter, locator })
}

export const start = (emitter, config, locator: Locator) => async ({ open, install, watcher }) => {
  const { scenarioUuid } = config

  try {
    await prepare(emitter, config, locator)({
      install,
      watchMode: watcher
    })

    const { scenarioRoot, buildViewsDir } = locator
    /* start local development server */
    const integrationHost = await startLocalDevelopmentServer(emitter, config, locator)

    ensureSetupComponents(emitter, locator)

    emitter.emit('start:watchers')
    if (watcher) {
      fs.watchFile(locator.authConfigPath, { persistent: true, interval: 250 }, () =>
        ensureSetupComponents(emitter, locator)
      )
    }

    /* Start bearer transpiler phase */
    const BEARER = 'bearer-transpiler'
    const options = [watcher ? null : '--no-watcher']

    // Build env for sub commands
    const envVariables = {
      ...process.env,
      BEARER_SCENARIO_TAG_NAME: 'localhost',
      BEARER_SCENARIO_ID: scenarioUuid,
      BEARER_INTEGRATION_HOST: integrationHost,
      BEARER_AUTHORIZATION_HOST: integrationHost
    }

    const bearerTranspiler = spawn(
      'node',
      [path.join(__dirname, '..', 'startTranspiler.js'), options].filter(el => el),
      {
        cwd: scenarioRoot,
        env: envVariables,
        stdio: ['pipe', 'pipe', 'pipe', 'ipc']
      }
    )
    bearerTranspiler.stdout.on('data', childProcessStdout(emitter, BEARER))
    bearerTranspiler.stderr.on('data', childProcessStderr(emitter, BEARER))
    bearerTranspiler.on('close', childProcessClose(emitter, BEARER))

    if (watcher) {
      const tsxWatcher = chokidar.watch('**/*.tsx', {
        ignored: /(^|[\/\\])\../,
        cwd: locator.srcViewsDir,
        ignoreInitial: true,
        persistent: true,
        followSymlinks: false
      })
      tsxWatcher.on('add', () => bearerTranspiler.send('refresh'))
      tsxWatcher.on('unlink', () => bearerTranspiler.send('refresh'))
      tsxWatcher.on('error', error => emitter.emit('start:watchers:componentError', { error }))

      bearerTranspiler.on('message', ({ event }) => {
        if (event === 'transpiler:initialized') {
          /* Start stencil */
          const args = config.isYarnInstalled ? ['start'] : ['run', 'start']
          if (!open) {
            args.push('--no-open')
          }
          const stencil = spawn(config.command, args, {
            cwd: buildViewsDir,
            env: envVariables
          })

          const STENCIL = 'stencil'

          stencil.stdout.on('data', childProcessStdout(emitter, STENCIL))
          stencil.stderr.on('data', childProcessStderr(emitter, STENCIL))
          stencil.on('close', childProcessClose(emitter, STENCIL))
        }
      })
    }
  } catch (e) {
    emitter.emit('start:failed', { error: e })
  }
}

export function useWith(program, emitter, config, locator: Locator) {
  program
    .command('start')
    .description(
      `Start local development server.
  $ bearer start
`
    )
    .option('--no-open', 'Do not open web browser')
    .option('--no-install', 'Do not run yarn|npm install')
    .option('--no-watcher', 'Run transpiler only once')
    .action(start(emitter, config, locator))
}

/**
 * Logger helpers
 */

function childProcessStdout(emitter, name) {
  return data => {
    emitter.emit('start:watchers:stdout', { name, data })
  }
}

function childProcessStderr(emitter, name) {
  return data => {
    emitter.emit('start:watchers:stderr', { name, data })
  }
}

function childProcessClose(emitter, name) {
  return code => {
    emitter.emit('start:watchers:close', { name, code })
  }
}

function createEvenIfItExists(target, sourcePath): void {
  try {
    fs.symlinkSync(target, sourcePath)
  } catch (e) {
    if (e.code !== 'EEXIST') {
      throw e
    }
  }
}
