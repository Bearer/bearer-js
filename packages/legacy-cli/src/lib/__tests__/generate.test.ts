import * as intents from '@bearer/intents'

import { getComponentVars, getIntentVars } from '../commands/generateCommand'
const TestingValues = ['spongebob', 'SpongeBob', 'spongeBob', 'sponge_bob', 'sponge-bob']
const AuthTypes = ['oauth2', 'apiKey', 'noAuth', 'basicAuth']

describe('Generate command', () => {
  describe('get components variables', () => {
    TestingValues.forEach(value => {
      AuthTypes.forEach(authType => {
        describe(authType, () => {
          describe(value, () => {
            it('formats variables correctly', () => {
              expect(getComponentVars(value, { authType })).toMatchSnapshot()
            })
          })
        })
      })
    })
  })

  describe('get intents variables', () => {
    Object.keys(intents).forEach(intentType => {
      AuthTypes.forEach(authType => {
        describe(authType, () => {
          describe(intentType, () => {
            TestingValues.forEach(value => {
              describe(value, () => {
                it('formats variables correctly', () => {
                  expect(getIntentVars(value, intentType, { authType })).toMatchSnapshot()
                })
              })
            })
          })
        })
      })
    })
  })
})
