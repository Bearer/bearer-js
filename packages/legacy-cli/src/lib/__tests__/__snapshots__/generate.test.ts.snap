// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate command get components variables SpongeBob formats variables correctly 1`] = `
Object {
  "componentName": "SpongeBob",
  "componentTagName": "sponge-bob",
  "fileName": "SpongeBob",
  "groupName": "sponge-bob",
}
`;

exports[`Generate command get components variables sponge_bob formats variables correctly 1`] = `
Object {
  "componentName": "SpongeBob",
  "componentTagName": "sponge-bob",
  "fileName": "sponge_bob",
  "groupName": "sponge-bob",
}
`;

exports[`Generate command get components variables sponge-bob formats variables correctly 1`] = `
Object {
  "componentName": "SpongeBob",
  "componentTagName": "sponge-bob",
  "fileName": "sponge-bob",
  "groupName": "sponge-bob",
}
`;

exports[`Generate command get components variables spongeBob formats variables correctly 1`] = `
Object {
  "componentName": "SpongeBob",
  "componentTagName": "sponge-bob",
  "fileName": "spongeBob",
  "groupName": "sponge-bob",
}
`;

exports[`Generate command get components variables spongebob formats variables correctly 1`] = `
Object {
  "componentName": "Spongebob",
  "componentTagName": "spongebob",
  "fileName": "spongebob",
  "groupName": "spongebob",
}
`;

exports[`Generate command get intents variables apiKey FetchData formats variables correctly 1`] = `
Object {
  "actionExample": "
  static action(context: TapiKeyContext, params: any, body: any, callback: TFetchDataCallback) {
    //... your code goes here
    // use the client defined in client.ts to fetch real object like that:
    // Client(context.authAccess.apiKey).get('/people')
    //   .then(({ data }) => {
    //     callback({ data })
    //   })
    //   .catch((error) => {
    //     callback({ error: error.toString() })
    //   })
    callback({ data: []})
  }
  ",
  "authType": "apiKey",
  "callbackType": "TFetchDataCallback",
  "intentName": "test",
  "intentType": "FetchData",
}
`;

exports[`Generate command get intents variables apiKey Intent formats variables correctly 1`] = `
Object {
  "actionExample": undefined,
  "authType": "apiKey",
  "callbackType": "TIntentCallback",
  "intentName": "test",
  "intentType": "Intent",
}
`;

exports[`Generate command get intents variables apiKey RetrieveState formats variables correctly 1`] = `
Object {
  "actionExample": "
  static action(_context: TapiKeyContext, _params: any, state, callback: TRetrieveStateCallback) {
    callback({ data: state })
  }
  ",
  "authType": "apiKey",
  "callbackType": "TRetrieveStateCallback",
  "intentName": "test",
  "intentType": "RetrieveState",
}
`;

exports[`Generate command get intents variables apiKey SaveState formats variables correctly 1`] = `
Object {
  "actionExample": "
  static action(
    _context: TapiKeyContext,
    _params: any,
    body: any,
    state: any,
    callback: TSaveStateCallback
  ): void {
    const { item: { name } } = body
    const { items = [] }: any = state
    const newItem: any = { name }

    callback({
      state: {
        ...state,
        items: [...items, newItem]
      },
      data: [...items, newItem]
    })
  }
  ",
  "authType": "apiKey",
  "callbackType": "TSaveStateCallback",
  "intentName": "test",
  "intentType": "SaveState",
}
`;

exports[`Generate command get intents variables basicAuth FetchData formats variables correctly 1`] = `
Object {
  "actionExample": "
  static action(context: TbasicAuthContext, params: any, body: any, callback: TFetchDataCallback) {
    //... your code goes here
    // use the client defined in client.ts to fetch real object like that:
    // Client(
    //   context.authAccess.username,
    //   context.authAccess.password
    // ).get('/people').then(({ data }) => {
    //     callback({ data })
    //   })
    //   .catch((error) => {
    //     callback({ error: error.toString() })
    //   })
    callback({ data: []})
  }
  ",
  "authType": "basicAuth",
  "callbackType": "TFetchDataCallback",
  "intentName": "test",
  "intentType": "FetchData",
}
`;

exports[`Generate command get intents variables basicAuth Intent formats variables correctly 1`] = `
Object {
  "actionExample": undefined,
  "authType": "basicAuth",
  "callbackType": "TIntentCallback",
  "intentName": "test",
  "intentType": "Intent",
}
`;

exports[`Generate command get intents variables basicAuth RetrieveState formats variables correctly 1`] = `
Object {
  "actionExample": "
  static action(_context: TbasicAuthContext, _params: any, state, callback: TRetrieveStateCallback) {
    callback({ data: state })
  }
  ",
  "authType": "basicAuth",
  "callbackType": "TRetrieveStateCallback",
  "intentName": "test",
  "intentType": "RetrieveState",
}
`;

exports[`Generate command get intents variables basicAuth SaveState formats variables correctly 1`] = `
Object {
  "actionExample": "
  static action(
    _context: TbasicAuthContext,
    _params: any,
    body: any,
    state: any,
    callback: TSaveStateCallback
  ): void {
    const { item: { name } } = body
    const { items = [] }: any = state
    const newItem: any = { name }

    callback({
      state: {
        ...state,
        items: [...items, newItem]
      },
      data: [...items, newItem]
    })
  }
  ",
  "authType": "basicAuth",
  "callbackType": "TSaveStateCallback",
  "intentName": "test",
  "intentType": "SaveState",
}
`;

exports[`Generate command get intents variables noAuth FetchData formats variables correctly 1`] = `
Object {
  "actionExample": "
  static action(context: TnoAuthContext, params: any, body: any, callback: TFetchDataCallback) {
    //... your code goes here
    callback({ data: []})
  }",
  "authType": "noAuth",
  "callbackType": "TFetchDataCallback",
  "intentName": "test",
  "intentType": "FetchData",
}
`;

exports[`Generate command get intents variables noAuth Intent formats variables correctly 1`] = `
Object {
  "actionExample": undefined,
  "authType": "noAuth",
  "callbackType": "TIntentCallback",
  "intentName": "test",
  "intentType": "Intent",
}
`;

exports[`Generate command get intents variables noAuth RetrieveState formats variables correctly 1`] = `
Object {
  "actionExample": "
  static action(_context: TnoAuthContext, _params: any, state, callback: TRetrieveStateCallback) {
    callback({ data: state })
  }
  ",
  "authType": "noAuth",
  "callbackType": "TRetrieveStateCallback",
  "intentName": "test",
  "intentType": "RetrieveState",
}
`;

exports[`Generate command get intents variables noAuth SaveState formats variables correctly 1`] = `
Object {
  "actionExample": "
  static action(
    _context: TnoAuthContext,
    _params: any,
    body: any,
    state: any,
    callback: TSaveStateCallback
  ): void {
    const { item: { name } } = body
    const { items = [] }: any = state
    const newItem: any = { name }

    callback({
      state: {
        ...state,
        items: [...items, newItem]
      },
      data: [...items, newItem]
    })
  }
  ",
  "authType": "noAuth",
  "callbackType": "TSaveStateCallback",
  "intentName": "test",
  "intentType": "SaveState",
}
`;

exports[`Generate command get intents variables oauth2 FetchData formats variables correctly 1`] = `
Object {
  "actionExample": "
  static action(context: Toauth2Context, params: any, body: any, callback: TFetchDataCallback) {
    //... your code goes here
    // use the client defined in client.ts to fetch real object like that:
    // Client(context.authAccess.accessToken).get('/people').then(({ data }) => {
    //     callback({ data })
    //   })
    //   .catch((error) => {
    //     callback({ error: error.toString() })
    //   })
    callback({ data: []})
  }",
  "authType": "oauth2",
  "callbackType": "TFetchDataCallback",
  "intentName": "test",
  "intentType": "FetchData",
}
`;

exports[`Generate command get intents variables oauth2 Intent formats variables correctly 1`] = `
Object {
  "actionExample": undefined,
  "authType": "oauth2",
  "callbackType": "TIntentCallback",
  "intentName": "test",
  "intentType": "Intent",
}
`;

exports[`Generate command get intents variables oauth2 RetrieveState formats variables correctly 1`] = `
Object {
  "actionExample": "
  static action(_context: Toauth2Context, _params: any, state, callback: TRetrieveStateCallback) {
    callback({ data: state })
  }
  ",
  "authType": "oauth2",
  "callbackType": "TRetrieveStateCallback",
  "intentName": "test",
  "intentType": "RetrieveState",
}
`;

exports[`Generate command get intents variables oauth2 SaveState formats variables correctly 1`] = `
Object {
  "actionExample": "
  static action(
    _context: Toauth2Context,
    _params: any,
    body: any,
    state: any,
    callback: TSaveStateCallback
  ): void {
    const { item: { name } } = body
    const { items = [] }: any = state
    const newItem: any = { name }

    callback({
      state: {
        ...state,
        items: [...items, newItem]
      },
      data: [...items, newItem]
    })
  }
  ",
  "authType": "oauth2",
  "callbackType": "TSaveStateCallback",
  "intentName": "test",
  "intentType": "SaveState",
}
`;
